# Data Preprocessing
This notebook consists of the following sections 
* About the Dataset
* Descovering Structure of the Data
* Data Transformation
* Data Cleaning
* Spliting & Exporting Data

## About the Dataset

The Used Cars in the Jordanian Second-Hand Market Dataset is a web-scraped dataset sourced from opensooq.com, an e-commerce platform used across several Middle Eastern countries, including Jordan. It supports both Customer-to-Customer (C2C) and Business-to-Customer (B2C) transactions for various products, including used cars.

### Dataset Documentation

In [1]:
#importing requried lirbraries 
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
#creating dictionary for storing meta data
doc_df = {
    "URL": {
        "Description": "Link to the car listing(Used as unique identfier).",
        "Value Range / Examples": "Example: 'https://jo.opensooq.com/en/search/257873181'"
    },
    "Scraped_Time": {
        "Description": "The timestamp when the data was scraped.",
        "Value Range / Examples": "Format: 'YYYY-MM-DD HH:MM:SS'"
    },
    "Condition": {
        "Description": "The condition of the car (New/Used).",
        "Value Range / Examples": ["New", "Used"]
    },
    "Car Make": {
        "Description": "The manufacturer of the car.",
        "Value Range / Examples": ["Toyota", "Hyundai", "Ford", "BMW"]
    },
    "Model": {
        "Description": "The specific model of the car.",
        "Value Range / Examples": ["Corolla", "Civic", "Camry", "X5"]
    },
    "Trim": {
        "Description": "The trim level or variant of the car.",
        "Value Range / Examples": ["SE", "Limited", "Sport", "LX"]
    },
    "Year": {
        "Description": "The manufacturing year of the car.",
        "Value Range / Examples": "Older than 1970 - 2025"
    },
    "Kilometers": {
        "Description": "The mileage range of the car.",
        "Value Range / Examples": "0 - 200,000+ km"
    },
    "Body Type": {
        "Description": "The body style of the car.",
        "Value Range / Examples": ["Sedan", "SUV", "Pickup", "Hatchback","Truck",'HatchBack', 'Coupe', 'Bus - Van','Convertible']
    },
    "Number of Seats": {
        "Description": "The number of seats in the car.",
        "Value Range / Examples": ["2 - more than 9"]
    },
    "Fuel": {
        "Description": "The type of fuel the car uses.",
        "Value Range / Examples":  ['Gasoline','Electric', 'Diesel', 'Hybrid', 'Plug-in - Hybrid', 'Mild Hybrid']
    },
    "Transmission": {
        "Description": "The type of transmission.",
        "Value Range / Examples": ["Automatic", "Manual"]
    },
    "Engine Size (cc)": {
        "Description": "The engine capacity in cubic centimeters.",
        "Value Range / Examples": "500 - 7000+ cc"
    },
    "Exterior Color": {
        "Description": "The color of the car's exterior.",
        "Value Range / Examples": ["Black", "White", "Red", "Blue", "etc."]
    },
    "Interior Color": {
        "Description": "The color of the car's interior.",
        "Value Range / Examples": ["Beige", "Black", "Gray", "Brown"]
    },
    "Regional Specs": {
        "Description": "Specifications based on region.",
        "Value Range / Examples":  ['Korean Specs', 'American Specs', 'Japanese Specs' ,'Other Specs','European Specs', 'GCC Specs']
    },
    "Car License": {
        "Description": "Details about the car's licensing status.",
        "Value Range / Examples":  ['Licensed','Not Licensed']
    },
    "Insurance": {
        "Description": "Whether the car has insurance or not.",
        "Value Range / Examples": ['Compulsory Insurance', 'Comprehensive Insurance', 'Not Insured']
    },
    "Car Customs": {
        "Description": "Indicates if the car has cleared customs.",
        "Value Range / Examples": ['With Customs','Without Customs']
    },
    "Body Condition": {
        "Description": "The condition of the car's body.",
        "Value Range / Examples": ['Excellent with no defects', 'Other', 'Good (body only has minor blemishes)', 'Fair (body needs work)', 'Poor (severe body damages)']
    },
    "Paint": {
        "Description": "The condition of the car's paint.",
        "Value Range / Examples": ["Original", "Repainted", "Partial Repaint"]
    },
    "Payment Method": {
        "Description": "The available payment options.",
        "Value Range / Examples": ["Cash", "Installments", 'Cash or Installments']
    },
    "City": {
        "Description": "The city where the car is listed.",
        "Value Range / Examples":  ['Irbid', 'Amman', 'Ajloun', 'Mafraq', 'Zarqa', 'Aqaba', 'Madaba', 'Salt', 'Al Karak','Jerash','Ramtha',"Ma'an",'Tafila','Jordan Valley']
    },
    "Neighborhood": {
        "Description": "The specific area within the city.",
        "Value Range / Examples": ['Al Quds Street', 'Abu Alanda', 'Tabarboor']
    },
    "Category": {
        "Description": "General category of the listing.",
        "Value Range / Examples": ["Cars and Bikes"]
    },
    "Subcategory": {
        "Description": "More specific category.",
        "Value Range / Examples": ["Cars for Sale", "Classic Cars"]
    },
    "Interior Options": {
        "Description": "Features related to the car's interior.",
        "Value Range / Examples": ['Alarm System', 'AUX / USB Input', 'Airbags', 'Electric Windows']
    },
    "Exterior Options": {
        "Description": "Features related to the car's exterior.",
        "Value Range / Examples": ['Daytime Running Lights', 'Spare Tyre', 'Rim Wheels', 'Rear Sensors']
    },
    "Technology Options": {
        "Description": "Technology-related features.",
        "Value Range / Examples": ["Navigation", "360° Camera", "Touch Screen"]
    },
    "Price": {
        "Description": "The listed price of the car, including currency.",
        "Value Range / Examples": "550 - 1000,000 JOD"
    },
    "Description_Score": {
        "Description": "Description or rating of the car's condition.",
        "Value Range / Examples": "4-7 (1 = Poor, 7 = Excellent)"
    },
    "Battery Capacity": {
        "Description": "For electric vehicles, the battery capacity range.",
        "Value Range / Examples": ["less than 50kwh - more than 100,000kwh"]
    },
    "Battery Range": {
        "Description": "The estimated driving range on a full charge.",
        "Value Range / Examples": "less than 100 - more than 500 km"
    }
}
#converting dictionary to table
doc_df = {
    "Column Name": list(doc_df.keys()),
    "Description": [doc_df[col]["Description"] for col in doc_df],
    "Value Range / Examples": [doc_df[col]["Value Range / Examples"] for col in doc_df]
}
#converting 'doc_df' from dictionary to dataframe
doc_df = pd.DataFrame(doc_df)

In [3]:
#printing out the table
from IPython.display import display
display(doc_df)


Column Name                                        Description  \
0                  URL  Link to the car listing(Used as unique identfi...   
1         Scraped_Time           The timestamp when the data was scraped.   
2            Condition               The condition of the car (New/Used).   
3             Car Make                       The manufacturer of the car.   
4                Model                     The specific model of the car.   
5                 Trim              The trim level or variant of the car.   
6                 Year                 The manufacturing year of the car.   
7           Kilometers                      The mileage range of the car.   
8            Body Type                         The body style of the car.   
9      Number of Seats                    The number of seats in the car.   
10                Fuel                     The type of fuel the car uses.   
11        Transmission                          The type of transmission.   
12    Engine Size (cc)          The engine capacity in cubic centimeters.   
13      Exterior Color                   The color of the car's exterior.   
14      Interior Color                   The color of the car's interior.   
15      Regional Specs                    Specifications based on region.   
16         Car License          Details about the car's licensing status.   
17           Insurance              Whether the car has insurance or not.   
18         Car Customs          Indicates if the car has cleared customs.   
19      Body Condition                   The condition of the car's body.   
20               Paint                  The condition of the car's paint.   
21      Payment Method                     The available payment options.   
22                City                  The city where the car is listed.   
23        Neighborhood                 The specific area within the city.   
24            Category                   General category of the listing.   
25         Subcategory                            More specific category.   
26    Interior Options            Features related to the car's interior.   
27    Exterior Options            Features related to the car's exterior.   
28  Technology Options                       Technology-related features.   
29               Price   The listed price of the car, including currency.   
30   Description_Score      Description or rating of the car's condition.   
31    Battery Capacity  For electric vehicles, the battery capacity ra...   
32       Battery Range      The estimated driving range on a full charge.   

                               Value Range / Examples  
0   Example: 'https://jo.opensooq.com/en/search/25...  
1                       Format: 'YYYY-MM-DD HH:MM:SS'  
2                                         [New, Used]  
3                        [Toyota, Hyundai, Ford, BMW]  
4                         [Corolla, Civic, Camry, X5]  
5                            [SE, Limited, Sport, LX]  
6                              Older than 1970 - 2025  
7                                     0 - 200,000+ km  
8   [Sedan, SUV, Pickup, Hatchback, Truck, HatchBa...  
9                                   [2 - more than 9]  
10  [Gasoline, Electric, Diesel, Hybrid, Plug-in -...  
11                                [Automatic, Manual]  
12                                     500 - 7000+ cc  
13                    [Black, White, Red, Blue, etc.]  
14                        [Beige, Black, Gray, Brown]  
15  [Korean Specs, American Specs, Japanese Specs,...  
16                           [Licensed, Not Licensed]  
17  [Compulsory Insurance, Comprehensive Insurance...  
18                    [With Customs, Without Customs]  
19  [Excellent with no defects, Other, Good (body ...  
20             [Original, Repainted, Partial Repaint]  
21         [Cash, Installments, Cash or Installments]  
22  [Irbid, Amman, Ajloun, Mafraq, Zarqa, Aqaba, M...  
23            [Al Quds Street, Abu Alanda, Tabar

## Descovering Structure of the Data

In [4]:
#Changing the view settings of dataframe 
pd.set_option('display.max_rows', None)  # Show all rows
pd.set_option('display.max_columns', None)  # Show all columns

In [5]:
#reading dataset
df = pd.read_csv("D:/mohammed/UNI/Dynamic Pricing for Used Cars in Jordan/Data/car_data.csv")

In [6]:
#view a sample of dataset of size 10
df.sample(10)

URL         Scraped_Time  \
1519   https://jo.opensooq.com/en/search/258113211  2025-02-06 22:37:03   
9440   https://jo.opensooq.com/en/search/257564961  2025-02-07 12:56:39   
1245   https://jo.opensooq.com/en/search/258662917  2025-02-06 22:32:42   
11461  https://jo.opensooq.com/en/search/257766593  2025-02-07 14:17:51   
4423   https://jo.opensooq.com/en/search/258366101  2025-02-07 01:45:15   
10781  https://jo.opensooq.com/en/search/257363989  2025-02-07 13:25:48   
10218  https://jo.opensooq.com/en/search/257468785  2025-02-07 13:08:13   
7663   https://jo.opensooq.com/en/search/257935979  2025-02-07 11:14:25   
6958   https://jo.opensooq.com/en/search/258047731  2025-02-07 11:03:50   
3382   https://jo.opensooq.com/en/search/258486131  2025-02-07 00:43:38   

      Condition    Car Make    Model      Trim  Year         Kilometers  \
1519        New        Neta    Other       NaN  2023                  0   
9440       Used      Nissan     Leaf         S  2013    90,000 - 99,999   
1245       Used         Kia   Optima        EX  2016    80,000 - 89,999   
11461      Used        Ford   Fusion        SE  2013           +200,000   
4423       Used        Opel   Vectra       NaN  1992  160,000 - 169,999   
10781      Used        Opel   Kadett       NaN  1990           +200,000   
10218      Used      Toyota    Hilux       GLX  2022      1,000 - 9,999   
7663       Used  Volkswagen     Bora    E-Bora  2021    40,000 - 49,999   
6958       Used  Volkswagen  Touareg  Standard  2006  160,000 - 169,999   
3382       Used       Honda    Civic       LXi  2006           +200,000   

       Body Type Number of Seats      Fuel Transmission  Engine Size (cc)  \
1519   HatchBack               5  Electric    Automatic               NaN   
9440   HatchBack               5  Electric    Automatic               NaN   
1245       Sedan               5    Hybrid    Automatic  2,000 - 2,999 cc   
11461      Sedan               5    Hybrid    Automatic  2,000 - 2,999 cc   
4423       Sedan               5  Gasoline       Manual               NaN   
10781  HatchBack               5  Gasoline       Manual  1,000 - 1,999 cc   
10218     PickUp               4    Diesel    Automatic  2,000 - 2,999 cc   
7663       Sedan               5  Electric    Automatic               NaN   
6958         SUV               5  Gasoline    Automatic  3,000 - 3,999 cc   
3382       Sedan               5  Gasoline    Automatic        0 - 499 cc   

      Exterior Color Interior Color  Regional Specs Car License  \
1519           Green          Black     Other Specs    Licensed   
9440           White          Black  Japanese Specs    Licensed   
1245           Black          Beige  American Specs    Licensed   
11461          Brown          Black  American Specs    Licensed   
4423        Burgundy     Nardo Grey  European Specs    Licensed   
10781      Navy Blue      Navy Blue  European Specs    Licensed   
10218          Black            Red       GCC Specs    Licensed   
7663           White          Black  European Specs    Licensed   
6958     Olive Green    Olive Green  European Specs    Licensed   
3382            Blue           Grey  Japanese Specs    Licensed   

                  Insurance   Car Customs             Body Condition  \
1519   Compulsory Insurance  With Customs                        NaN   
9440   Compulsory Insurance  With Customs                        NaN   
1245   Compulsory Insurance  With Customs  Excellent with no defects   
11461  Compulsory Insurance  With Customs                      Other   
4423   Compulsory Insurance  With Customs  Excellent with no defects   
10781  Compulsory Insurance  With Customs  Excellent with no defects   
10218  Compulsory Insurance  With Customs  Excellent with no defects   
7663   Compulsory Insurance  With Customs                      Other   
6958   Compulsory Insurance  With Customs                      Other   
3382   Compulsory Insurance  With Customs  Excellent with no defects   

                

In [7]:
#storing the number of columns and rows
num_col = df.shape[1] #number of columns
num_rows = df.shape[0] #number of rows
print(f"Number of Attributes",num_col)
print(f"Number of Entries",num_rows)

Number of Attributes 33
Number of Entries 12003


The dataset contains *33 Attributes* with *12003 Entries* all of *Object* data type, they could be text, date/time, categorical and numerical; so as a convention they may be converted to the proper data type later in the transformation stage  

In [8]:
#get basic info about dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12003 entries, 0 to 12002
Data columns (total 33 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   URL                 12003 non-null  object
 1   Scraped_Time        12003 non-null  object
 2   Condition           11991 non-null  object
 3   Car Make            11990 non-null  object
 4   Model               11990 non-null  object
 5   Trim                10293 non-null  object
 6   Year                11992 non-null  object
 7   Kilometers          11984 non-null  object
 8   Body Type           11991 non-null  object
 9   Number of Seats     11986 non-null  object
 10  Fuel                11984 non-null  object
 11  Transmission        11984 non-null  object
 12  Engine Size (cc)    9402 non-null   object
 13  Exterior Color      11988 non-null  object
 14  Interior Color      11984 non-null  object
 15  Regional Specs      11991 non-null  object
 16  Car License         11

In [9]:
#unique values in each attribute
for attr in df.columns:
    print(attr,"\n",df[attr].unique())

URL 
 ['https://jo.opensooq.com/en/search/258270309'
 'https://jo.opensooq.com/en/search/258415399'
 'https://jo.opensooq.com/en/search/257669123' ...
 'https://jo.opensooq.com/en/search/258602761'
 'https://jo.opensooq.com/en/search/258955829'
 'https://jo.opensooq.com/en/search/258851377']
Scraped_Time 
 ['2025-02-06 20:14:39' '2025-02-06 20:16:26' '2025-02-06 20:16:43' ...
 '2025-02-12 22:17:54' '2025-02-12 22:17:57' '2025-02-12 22:17:58']
Condition 
 ['Used' 'New' nan]
Car Make 
 ['Hyundai' 'Tesla' 'Mitsubishi' 'Toyota' 'Kia' 'Lexus' 'Isuzu' 'Ford'
 'Volkswagen' 'Chevrolet' 'Mercedes Benz' 'Cadillac' 'BMW' 'Land Rover'
 'Nissan' 'Honda' 'GMC' 'Opel' 'Mazda' 'Samsung' 'Fiat' 'Audi' 'Suzuki'
 'Leapmotor' 'Peugeot' 'Jeep' 'Lincoln' 'Renault' 'Dodge' 'MG' 'Citroen'
 'Saab' 'Changan' 'Neta' 'BYD' 'Infiniti' 'Daewoo' 'Skywell' 'Chery'
 'Daihatsu' 'Genesis' 'Avatar' 'Porsche' 'Dongfeng' 'Jaguar' 'Volvo'
 'Hongqi' 'Lada' nan 'GAC' 'Abarath' 'Haval' 'Skoda' 'Chrysler' 'Bestune'
 'Mercury' '

It's important to know that not all missing values are orginated from typical reasons for missing values suach as; user does not enter the data or the bot that did not scrape data probably, where the most of the missing values are orginated from the different types of engines, for example Elctric cars has it's own features that differs from non-elctric cars

In [10]:
#checking for missing values
df.isnull().sum()

URL                      0
Scraped_Time             0
Condition               12
Car Make                13
Model                   13
Trim                  1710
Year                    11
Kilometers              19
Body Type               12
Number of Seats         17
Fuel                    19
Transmission            19
Engine Size (cc)      2601
Exterior Color          15
Interior Color          19
Regional Specs          12
Car License             19
Insurance               19
Car Customs             19
Body Condition         748
Paint                  778
Payment Method          19
City                     6
Neighborhood             6
Category                 6
Subcategory              6
Interior Options        19
Exterior Options        19
Technology Options      20
Price                  475
Description_Score     7236
Battery Capacity      9903
Battery Range         9903
dtype: int64

URL' atrribute used as unique identifier, where each link cotains unique id for each car ad, hence URL's are not redundant, and if there is a redundant URL it shuold be removed. As shown in Descovering Structure of the Data section the number of duplicated rows in general are 6, while the number of dulpicates by URL are 76 and that is another reasons why using URL to handle duplicates

In [11]:
#checking for duplicates for all dataset
df.duplicated().sum()

6

In [12]:
#checking for duplicates based on 'URL' atrribute
df['URL'].duplicated().sum()

76

## Data Transformation

In [13]:
#creating new dataset for cleaning 
clean_df = df.copy()

### Handling Irrelevant Atrtibutes 

Attribute with only single value across all data points does not have any significant added value for analysis or modeling, therefore they are dropped

In [14]:
# printing unique values in 'Category' attribute
clean_df['Category'].unique()

array(['Cars and Bikes', nan], dtype=object)

In [15]:
# printing unique values in 'Subcategory' attribute
clean_df['Subcategory'].unique()

array(['Cars For Sale', nan], dtype=object)

In [16]:
clean_df['Description_Score'].unique()

array(['فحص كامل', nan, '4', '7', '2', '3', '1', '2013', '20062', '20264',
       '15', '5', '20123', '1500', '20182', '4074', '82', '60', '18002',
       '6', '867', '28', '1007', '20194', '18004', '20217', '20143',
       '20002', '400007', '2006', '20232', '20254', '3820007', '3600',
       '20073', '24'], dtype=object)

In [17]:
#dropping 'Subcategory' and 'Category'
clean_df.drop(columns=['Subcategory','Category','Description_Score'], inplace = True)

In [18]:
# shape of the data after dropping 'Subcategory' and 'Category'
clean_df.shape

(12003, 30)

### Converting Datatpyes

Converting columns to thair propper datatypes, as following:
* 'Scraped_Time' from object to date/time.
* Year from object to integer, where all values under 'Year' attribute are numrical unless one value 'Older than 1970' with count of 5 values, as solving this problem imputation of 'Older than 1970' with '1970' were performed, while the count of the imputed value is very small.
* converting 'Price' from object to integer; to convert price to integer extracting the price from currency using RegEx then converting it.

In [19]:
#convert 'Scraped_Time' from object to date/time
clean_df['Scraped_Time'] = pd.to_datetime(clean_df["Scraped_Time"])
clean_df['Scraped_Time'].dtype

dtype('<M8[ns]')

In [20]:
df[df['Year']== 'Older than 1970'].shape[0]

5

In [21]:
clean_df['Year'] = df['Year'].replace('Older than 1970', '1970')

In [22]:
clean_df[clean_df['Year']== 'Older than 1970'].shape[0]

0

In [23]:
#convert 'Year' from object to integer
clean_df["Year"] = clean_df["Year"].astype(float)
clean_df['Year'].dtype

dtype('float64')

In [24]:
#extracting the price from currency    
clean_df['Price'] = clean_df['Price'].str.replace(r'[^\d.]', '', regex=True)
clean_df['Price'].sample(5)

6099    34500
346     25800
4699     5500
8985     1200
6355     4700
Name: Price, dtype: object

In [25]:
#converting 'Price' from object to flaot
clean_df['Price'] = clean_df['Price'].astype(float)
clean_df['Price'].dtype

dtype('float64')

## Data Cleaning

### Handling Missing Value

Regarding missing values; not all missing values in each attribute are actually misssing, depending on the type of fuel of speceific car, for example Non-Elctric cars don't have 'Battery Capacity' and 'Battery Range', vice versa Elctric cars don't have 'Engine Size (cc)', handling missing values as following:
* Handling missing values in common attributes between two diffirent types of cars: Non-Elctric and Elctric cars, and handle them by imputation as needed
* Spliting dataset after cleaning into two main datasets based on cars type
* Due to the importance of the 'Price' attribute there is a section to handle it's missing values and outliers
* Records with more than or equal 15 missing values are considered as Corrupted rows, to handle them they were dropped
*  

In [26]:
#specify selected columns to be handleled either by dropping or imputing 
selected_col = clean_df.select_dtypes(include = ['object']).columns #storing "Object" datatype columns 
selected_col = set(selected_col) #convert from list to set
selected_col = selected_col -  {'Trim','Engine Size (cc)','Body Condition','Paint','Description_Score','Battery Capacity', 'Battery Range','URL','Year'} #select the specefic columns
print(selected_col)

{'Interior Options', 'Regional Specs', 'Fuel', 'Exterior Options', 'Kilometers', 'Body Type', 'Insurance', 'Model', 'Interior Color', 'Transmission', 'Payment Method', 'Technology Options', 'Car License', 'Car Customs', 'City', 'Exterior Color', 'Number of Seats', 'Neighborhood', 'Car Make', 'Condition'}


In [27]:
# records with 15 or more 
df[df.isna().sum(axis=1)>=15].shape[0]

19

In [28]:
#storing indices of corruoted rows that contains more than or equal to 15 missing values
corrupted_rows = clean_df[clean_df.isna().sum(axis=1)>=15].index

In [29]:
#dropping corrupted rows 
clean_df.drop(corrupted_rows,inplace = True)

In [30]:
clean_df.shape

(11984, 30)

In [31]:
clean_df.isna().sum()

URL                      0
Scraped_Time             0
Condition                0
Car Make                 3
Model                    3
Trim                  1691
Year                     0
Kilometers               0
Body Type                1
Number of Seats          2
Fuel                     0
Transmission             0
Engine Size (cc)      2582
Exterior Color           0
Interior Color           3
Regional Specs           0
Car License              0
Insurance                0
Car Customs              0
Body Condition         730
Paint                  760
Payment Method           0
City                     0
Neighborhood             0
Interior Options         0
Exterior Options         0
Technology Options       1
Price                  469
Battery Capacity      9884
Battery Range         9884
dtype: int64

In [32]:
#filling missing values with most frequency value(mode) in each atrribute
for col in selected_col: #iterating over selected columns 
    print(col,f"-->",clean_df[col].mode()[0]) #printing mode
    clean_df[col] = clean_df[col].fillna(clean_df[col].mode()[0]) #filling missing values

Interior Options --> AUX / USB Input
Regional Specs --> Korean Specs
Fuel --> Gasoline
Exterior Options --> Rim Wheels
Kilometers --> +200,000
Body Type --> Sedan
Insurance --> Compulsory Insurance
Model --> Avante
Interior Color --> Black
Transmission --> Automatic
Payment Method --> Cash Only
Technology Options --> Bluetooth
Car License --> Licensed
Car Customs --> With Customs
City --> Amman
Exterior Color --> White
Number of Seats --> 5
Neighborhood --> Al mantika Al Hurra
Car Make --> Hyundai
Condition --> Used


For *'Trim'* atrribute, not all cars have a specific *'Trim'* they just have *'Model'* value; which something depends on the manufacturer of the car, to handle missing values in *'Trim'* atrribute, the value of the *'Model'* of specefic car is imputed in the *'Trim'*; since *'Model'* is the most logical and does not have many missing values

In [33]:
#finding missing values in 'Trim' and impute them by 'Model' value
clean_df['Trim'] = clean_df['Trim'].fillna(clean_df[clean_df['Trim'].isna()].Model)

In [34]:
#before imputing 'Model' values in 'Trim'
df[df['URL']=='https://jo.opensooq.com/en/search/257301835']

URL         Scraped_Time  \
78  https://jo.opensooq.com/en/search/257301835  2025-02-06 20:19:02   

   Condition Car Make   Model Trim  Year       Kilometers Body Type  \
78      Used  Hyundai  Sonata  NaN  2018  70,000 - 79,999     Sedan   

   Number of Seats    Fuel Transmission  Engine Size (cc) Exterior Color  \
78               5  Hybrid    Automatic  1,000 - 1,999 cc          White   

   Interior Color Regional Specs Car License             Insurance  \
78          Brown   Korean Specs    Licensed  Compulsory Insurance   

     Car Customs             Body Condition           Paint Payment Method  \
78  With Customs  Excellent with no defects  Original Paint      Cash Only   

     City Neighborhood        Category    Subcategory  \
78  Amman  Sports City  Cars and Bikes  Cars For Sale   

                                     Interior Options  \
78  Electric Seat Control, Memory Seats, Steering ...   

                                     Exterior Options  \
78  Spare Tyre, Rim Wheels, Rear Sensors, Keyless ...   

                                   Technology Options       Price  \
78  Touch Screen, Voice Control, Media Screen, Bli...  18,500 JOD   

   Description_Score Battery Capacity Battery Range  
78                 7              NaN           NaN

In [35]:
#after imputing 'Model' values in 'Trim'
clean_df[clean_df['URL']=='https://jo.opensooq.com/en/search/257301835']

URL        Scraped_Time Condition  \
78  https://jo.opensooq.com/en/search/257301835 2025-02-06 20:19:02      Used   

   Car Make   Model    Trim    Year       Kilometers Body Type  \
78  Hyundai  Sonata  Sonata  2018.0  70,000 - 79,999     Sedan   

   Number of Seats    Fuel Transmission  Engine Size (cc) Exterior Color  \
78               5  Hybrid    Automatic  1,000 - 1,999 cc          White   

   Interior Color Regional Specs Car License             Insurance  \
78          Brown   Korean Specs    Licensed  Compulsory Insurance   

     Car Customs             Body Condition           Paint Payment Method  \
78  With Customs  Excellent with no defects  Original Paint      Cash Only   

     City Neighborhood                                   Interior Options  \
78  Amman  Sports City  Electric Seat Control, Memory Seats, Steering ...   

                                     Exterior Options  \
78  Spare Tyre, Rim Wheels, Rear Sensors, Keyless ...   

                                   Technology Options    Price  \
78  Touch Screen, Voice Control, Media Screen, Bli...  18500.0   

   Battery Capacity Battery Range  
78              NaN           NaN

It was noticed in the 'Body Condition' attribute 655 missing values from 676 have New 'Condition', from the domain view most of the cars of New condition have an Excellent body condition, therefore they were imputed with the value of ''Excellent with no defects', while the rest of missing values of 21 records are imputed with 'Fair (body needs work)' 

In [36]:
# total number of missing values in 'Body Condition'
clean_df['Body Condition'].isna().sum()

730

In [37]:
# number of missing values in 'Body Condition' with ['Condition'] == New
clean_df[(clean_df['Body Condition'].isna())&(clean_df['Condition']=='New')].shape[0]

708

In [38]:
# impute missing values in 'Body Condition' with condition of ['Condition']=='New' with of 'Excellent with no defects'
clean_df.loc[
    (clean_df['Body Condition'].isna()) & (clean_df['Condition'] == 'New'), 
    'Body Condition'] = clean_df['Body Condition'].fillna('Excellent with no defects')

In [39]:
# number of the rest of missing values in 'Body Condition'
clean_df['Body Condition'].isna().sum()

22

In [40]:
# impute missing values in 'Body Condition' with value of 'Fair (body needs work)'
clean_df.loc[
    (clean_df['Body Condition'].isna()), 
    'Body Condition'] = clean_df['Body Condition'].fillna('Fair (body needs work)')

In [41]:
clean_df['Body Condition'].isna().sum()

0

In 'Paint' attribute showing the origine of the car paint, where 702 are missing 658 are of 'Condition' equal to New, which makes more sense to be 'Original Paint', therefore they were imputed by 'Original Paint', while 44 missing values with 'Condition' equal to Used are imputed with the mode of 'Paint' attribute

In [42]:
clean_df['Paint'].isna().sum()

760

In [43]:
clean_df[(clean_df['Paint'].isna())&(clean_df['Condition']=='New')].shape[0]

712

In [44]:
clean_df.loc[
    (clean_df['Paint'].isna()) & (clean_df['Condition'] == 'New'), 
    'Paint'] = clean_df['Paint'].fillna('Original Paint')

In [45]:
clean_df['Paint'].isna().sum()

48

In [46]:
clean_df['Paint'].mode()

0    Original Paint
Name: Paint, dtype: object

In [47]:
clean_df['Paint'].fillna(clean_df['Paint'].mode()[0], inplace=True)

In [48]:
clean_df['Paint'].isna().sum()

0

In [49]:
clean_df[(clean_df['Engine Size (cc)'].isna())&(clean_df['Fuel']!='Electric')].shape[0]

508

In [50]:
# Define the condition for missing Engine Size
condition = clean_df['Fuel'].isin(['Gasoline', 'Diesel', 'Hybrid', 'Plug-in - Hybrid', 'Mild Hybrid']) & clean_df['Engine Size (cc)'].isna()

# Step 1: Fill using Model + Year mode
mode_values = clean_df.groupby(["Model", "Year"])['Engine Size (cc)'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)
clean_df.loc[condition, 'Engine Size (cc)'] = mode_values

# Step 2: Fill remaining NaNs using Model-only mode
still_missing = clean_df['Engine Size (cc)'].isna()
mode_values_model = clean_df.groupby("Model")['Engine Size (cc)'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)
clean_df.loc[still_missing, 'Engine Size (cc)'] = mode_values_model

# Step 3: Fill remaining NaNs with the overall mode
overall_mode = clean_df['Engine Size (cc)'].mode()
if not overall_mode.empty:
    clean_df['Engine Size (cc)'].fillna(overall_mode.iloc[0], inplace=True)

# Print remaining NaNs
remaining_missing = clean_df['Engine Size (cc)'].isna().sum()
print(f"Final missing values in 'Engine Size (cc)': {remaining_missing}")


Final missing values in 'Engine Size (cc)': 0


In [51]:
clean_df[(clean_df['Engine Size (cc)'].isna())&(clean_df['Fuel']!='Electric')].shape[0]

0

### Handling Outliers and Missing Values in 'Price' Attribute

First handling outliers then handling missing values as following: 
* Grouping cars that have simeller price ranges into groups by 'Model' and 'Year', where the most two significant factors to group cars by are 'Model' and 'Year'
* After grouping cars, detecting outliers by implementing IQR methode, then dropping outliers
*  Because of small groups and the most of prices in it are outliers, detecting ouliers by condition, where Year>=2015 and price<=10000 and Payment Methode == 'Cash or Installements' and 'Installements Only', then imputing these outliers by taking the mean of the specefic group
*  Like the previous step, also imputing outliers but thie time with the following condition: Year>=2010 and price<=3000 and Payment Methode == 'Cash Only'
*  Still there were 101 record needs handling manually 

In [52]:
backup_df = clean_df.copy()

In [53]:
def IQR_outliers(gr):
    """
    Identifies outliers in the 'Price' column using the Interquartile Range (IQR) method.

    Parameters:
    -----------
    gr : pandas.DataFrame
        A DataFrame or grouped DataFrame that contains a 'Price' column.

    Returns:
    --------
    pandas.DataFrame
        A subset of the original DataFrame containing only the outlier rows 
        where the 'Price' is below the lower bound or above the upper bound.

    Explanation:
    ------------
    The function follows these steps:
    1. Calculates the first quartile (Q1) and third quartile (Q3) of the 'Price' column.
    2. Computes the Interquartile Range (IQR) as Q3 - Q1.
    3. Defines the lower and upper bounds for detecting outliers:
       - Lower Bound = Q1 - 1.5 * IQR
       - Upper Bound = Q3 + 1.5 * IQR
    4. Filters and returns rows where 'Price' is outside these bounds.
    """

    Q1 = gr['Price'].quantile(0.25)  # First quartile (25th percentile)
    Q3 = gr['Price'].quantile(0.75)  # Third quartile (75th percentile)
    IQR = Q3 - Q1  # Compute the IQR
    lower_bound = Q1 - 1.5 * IQR  # Define lower bound
    upper_bound = Q3 + 1.5 * IQR  # Define upper bound

    # Return only the outlier rows
    return gr[(gr["Price"] < lower_bound) | (gr["Price"] > upper_bound)]


#### Dropping outliers

In [54]:
grouped = clean_df.groupby(["Model", "Year"])
grouped.ngroups

2961

In [55]:
# detecting outliers in each group
grouped = clean_df.groupby(["Model", "Year"])
dfs = []  # list for adding outliers 
for (model, year), group in grouped: # iterating over groups
    temp =IQR_outliers(group) # detecting outliers
    dfs.append(temp) 
outliers_df = pd.concat(dfs)

In [56]:
outliers_df.shape[0]

691

In [57]:
outliers_indices = outliers_df.index 

In [58]:
# dropping outliers
clean_df.drop(outliers_indices,inplace =True)

#### Imputing Outliers

In [59]:
clean_df[((clean_df['Payment Method']=='Cash or Installments')|(clean_df['Payment Method']=='Installments Only'))&(clean_df['Price']<=10000)&(clean_df['Year']>=2015)].shape[0]


458

In [60]:
grouped = clean_df.groupby(["Model", "Year"])

for (model, year), group in grouped:
    # Apply condition within each group to avoid reindexing issues
    condition = ((group['Payment Method'].isin(['Cash or Installments', 'Installments Only'])) & 
                 (group['Year'] >= 2015) & 
                 (group['Price'] <= 10000))

    filtered_group = group[condition]  # Apply condition correctly

    if not filtered_group.empty:  # Ensure there are matching rows
        indx = filtered_group.index  # Get indices of matching rows     
        mean_price = group['Price'].mean()  # Mean of the group

        # Update 'Price' for selected rows
        clean_df.loc[indx, 'Price'] = mean_price

print("Final shape of clean_df:", clean_df.shape)


Final shape of clean_df: (11293, 30)


In [61]:
clean_df[((clean_df['Payment Method']=='Cash or Installments')|(clean_df['Payment Method']=='Installments Only'))&(clean_df['Price']<=3000)&(clean_df['Year']>=2010)].shape[0]

46

In [62]:
indx = clean_df[((clean_df['Payment Method']=='Cash or Installments')|(clean_df['Payment Method']=='Installments Only'))&(clean_df['Price']<=3000)&(clean_df['Year']>=2010)].index

In [63]:
clean_df.drop(indx,inplace =True)

In [64]:
clean_df.shape

(11247, 30)

In [65]:
#number of outliers where clean_df['Payment Method']=='Cash Only')&(clean_df['Year']>=2010)
clean_df[(clean_df['Price']<=5000)&(clean_df['Payment Method']=='Cash Only')&(clean_df['Year']>=2010)].shape[0]

325

In [66]:
# storing indx of outliers
tmp_indx = clean_df[(clean_df['Price']<=5000)&(clean_df['Payment Method']=='Cash Only')&(clean_df['Year']>=2010)].index
# outliers before imputing
clean_df[(clean_df['Price']<=5000)&(clean_df['Payment Method']=='Cash Only')&(clean_df['Year']>=2010)]

URL        Scraped_Time  \
35     https://jo.opensooq.com/en/search/258749739 2025-02-06 20:17:39   
42     https://jo.opensooq.com/en/search/258749261 2025-02-06 20:17:48   
74     https://jo.opensooq.com/en/search/258699163 2025-02-06 20:18:55   
79     https://jo.opensooq.com/en/search/258746185 2025-02-06 20:19:03   
100    https://jo.opensooq.com/en/search/258745361 2025-02-06 20:19:29   
162    https://jo.opensooq.com/en/search/257102417 2025-02-06 21:20:32   
285    https://jo.opensooq.com/en/search/258737337 2025-02-06 21:23:17   
294    https://jo.opensooq.com/en/search/258512139 2025-02-06 21:23:24   
315    https://jo.opensooq.com/en/search/258736167 2025-02-06 21:23:45   
391    https://jo.opensooq.com/en/search/258733895 2025-02-06 21:24:54   
421    https://jo.opensooq.com/en/search/258732973 2025-02-06 21:25:47   
434    https://jo.opensooq.com/en/search/258732057 2025-02-06 21:26:18   
472    https://jo.opensooq.com/en/search/258729985 2025-02-06 21:27:33   
591    https://jo.opensooq.com/en/search/258672001 2025-02-06 21:32:42   
609    https://jo.opensooq.com/en/search/258717851 2025-02-06 21:33:03   
636    https://jo.opensooq.com/en/search/258314841 2025-02-06 21:33:37   
698    https://jo.opensooq.com/en/search/258112769 2025-02-06 21:35:08   
769    https://jo.opensooq.com/en/search/258655151 2025-02-06 21:36:30   
776    https://jo.opensooq.com/en/search/258704483 2025-02-06 21:36:37   
862    https://jo.opensooq.com/en/search/258696857 2025-02-06 22:26:33   
863    https://jo.opensooq.com/en/search/257705829 2025-02-06 22:26:34   
901    https://jo.opensooq.com/en/search/258644639 2025-02-06 22:27:09   
916    https://jo.opensooq.com/en/search/258693527 2025-02-06 22:27:23   
931    https://jo.opensooq.com/en/search/257557327 2025-02-06 22:27:39   
945    https://jo.opensooq.com/en/search/258691923 2025-02-06 22:27:52   
1060   https://jo.opensooq.com/en/search/258357595 2025-02-06 22:29:50   
1073   https://jo.opensooq.com/en/search/258683349 2025-02-06 22:30:04   
1126   https://jo.opensooq.com/en/search/258605725 2025-02-06 22:30:54   
1151   https://jo.opensooq.com/en/search/258674929 2025-02-06 22:31:17   
1303   https://jo.opensooq.com/en/search/258658227 2025-02-06 22:33:33   
1321   https://jo.opensooq.com/en/search/258656627 2025-02-06 22:33:48   
1350   https://jo.opensooq.com/en/search/258653859 2025-02-06 22:34:21   
1487   https://jo.opensooq.com/en/search/258635477 2025-02-06 22:36:33   
1495   https://jo.opensooq.com/en/search/258642739 2025-02-06 22:36:40   
1533   https://jo.opensooq.com/en/search/258639499 2025-02-06 22:37:16   
1562   https://jo.opensooq.com/en/search/258587347 2025-02-06 22:37:46   
1591   https://jo.opensooq.com/en/search/258636745 2025-02-06 22:38:13   
1634   https://jo.opensooq.com/en/search/258633525 2025-02-06 22:38:54   
1667   https://jo.opensooq.com/en/search/258583629 2025-02-06 22:39:25   
1668   https://jo.opensooq.com/en/search/258583651 2025-02-06 22:39:26   
1689   https://jo.opensooq.com/en/search/258577601 2025-02-06 22:39:45   
1696   https://jo.opensooq.com/en/search/258628481 2025-02-06 22:39:51   
1774   https://jo.opensooq.com/en/search/258621629 2025-02-06 22:41:03   
1803   https://jo.opensooq.com/en/search/258565287 2025-02-06 22:41:30   
1841   https://jo.opensooq.com/en/search/258162013 2025-02-06 22:42:04   
1855   https://jo.opensooq.com/en/search/258612209 2025-02-06 22:42:18   
1896   https://jo.opensooq.com/en/search/258610063 2025-02-06 22:42:56   
1933   https://jo.opensooq.com/en/search/258607265 2025-02-06 22:43:32   
1940   https://jo.opensooq.com/en/search/258604317 2025-02-06 22:43:38   
2017   https://jo.opensooq.com/en/search/257306655 2025-02-06 23:33:19   
2067   https://jo.opensooq.com/en/search/258596595 2025-02-06 23:34:02   
2120   https://jo.opensooq.com/en/search/258583659 2025-02-06 23:34:47   
2128   https://jo.opensooq.com/en/search/258590863 2025-02-06 23:34:53   
2132   https://jo.opensooq.com/en/search/258590219 

In [67]:
grouped = clean_df.groupby(["Model", "Year"])

for (model, year), group in grouped:
    # Apply condition within each group to avoid reindexing issues
    condition = ((group['Payment Method'].isin(['Cash Only'])) & 
                 (group['Year'] >= 2010) & 
                 (group['Price'] <= 3000))

    filtered_group = group[condition]  # Apply condition correctly

    if not filtered_group.empty:  # Ensure there are matching rows
        indx = filtered_group.index  # Get indices of matching rows     
        mean_price = group['Price'].mean()  # Mean of the group

        # Update 'Price' for selected rows
        clean_df.loc[indx, 'Price'] = mean_price

print("Final shape of clean_df:", clean_df.shape)

Final shape of clean_df: (11247, 30)


In [68]:
# after imputing
clean_df.loc[tmp_indx]

URL        Scraped_Time  \
35     https://jo.opensooq.com/en/search/258749739 2025-02-06 20:17:39   
42     https://jo.opensooq.com/en/search/258749261 2025-02-06 20:17:48   
74     https://jo.opensooq.com/en/search/258699163 2025-02-06 20:18:55   
79     https://jo.opensooq.com/en/search/258746185 2025-02-06 20:19:03   
100    https://jo.opensooq.com/en/search/258745361 2025-02-06 20:19:29   
162    https://jo.opensooq.com/en/search/257102417 2025-02-06 21:20:32   
285    https://jo.opensooq.com/en/search/258737337 2025-02-06 21:23:17   
294    https://jo.opensooq.com/en/search/258512139 2025-02-06 21:23:24   
315    https://jo.opensooq.com/en/search/258736167 2025-02-06 21:23:45   
391    https://jo.opensooq.com/en/search/258733895 2025-02-06 21:24:54   
421    https://jo.opensooq.com/en/search/258732973 2025-02-06 21:25:47   
434    https://jo.opensooq.com/en/search/258732057 2025-02-06 21:26:18   
472    https://jo.opensooq.com/en/search/258729985 2025-02-06 21:27:33   
591    https://jo.opensooq.com/en/search/258672001 2025-02-06 21:32:42   
609    https://jo.opensooq.com/en/search/258717851 2025-02-06 21:33:03   
636    https://jo.opensooq.com/en/search/258314841 2025-02-06 21:33:37   
698    https://jo.opensooq.com/en/search/258112769 2025-02-06 21:35:08   
769    https://jo.opensooq.com/en/search/258655151 2025-02-06 21:36:30   
776    https://jo.opensooq.com/en/search/258704483 2025-02-06 21:36:37   
862    https://jo.opensooq.com/en/search/258696857 2025-02-06 22:26:33   
863    https://jo.opensooq.com/en/search/257705829 2025-02-06 22:26:34   
901    https://jo.opensooq.com/en/search/258644639 2025-02-06 22:27:09   
916    https://jo.opensooq.com/en/search/258693527 2025-02-06 22:27:23   
931    https://jo.opensooq.com/en/search/257557327 2025-02-06 22:27:39   
945    https://jo.opensooq.com/en/search/258691923 2025-02-06 22:27:52   
1060   https://jo.opensooq.com/en/search/258357595 2025-02-06 22:29:50   
1073   https://jo.opensooq.com/en/search/258683349 2025-02-06 22:30:04   
1126   https://jo.opensooq.com/en/search/258605725 2025-02-06 22:30:54   
1151   https://jo.opensooq.com/en/search/258674929 2025-02-06 22:31:17   
1303   https://jo.opensooq.com/en/search/258658227 2025-02-06 22:33:33   
1321   https://jo.opensooq.com/en/search/258656627 2025-02-06 22:33:48   
1350   https://jo.opensooq.com/en/search/258653859 2025-02-06 22:34:21   
1487   https://jo.opensooq.com/en/search/258635477 2025-02-06 22:36:33   
1495   https://jo.opensooq.com/en/search/258642739 2025-02-06 22:36:40   
1533   https://jo.opensooq.com/en/search/258639499 2025-02-06 22:37:16   
1562   https://jo.opensooq.com/en/search/258587347 2025-02-06 22:37:46   
1591   https://jo.opensooq.com/en/search/258636745 2025-02-06 22:38:13   
1634   https://jo.opensooq.com/en/search/258633525 2025-02-06 22:38:54   
1667   https://jo.opensooq.com/en/search/258583629 2025-02-06 22:39:25   
1668   https://jo.opensooq.com/en/search/258583651 2025-02-06 22:39:26   
1689   https://jo.opensooq.com/en/search/258577601 2025-02-06 22:39:45   
1696   https://jo.opensooq.com/en/search/258628481 2025-02-06 22:39:51   
1774   https://jo.opensooq.com/en/search/258621629 2025-02-06 22:41:03   
1803   https://jo.opensooq.com/en/search/258565287 2025-02-06 22:41:30   
1841   https://jo.opensooq.com/en/search/258162013 2025-02-06 22:42:04   
1855   https://jo.opensooq.com/en/search/258612209 2025-02-06 22:42:18   
1896   https://jo.opensooq.com/en/search/258610063 2025-02-06 22:42:56   
1933   https://jo.opensooq.com/en/search/258607265 2025-02-06 22:43:32   
1940   https://jo.opensooq.com/en/search/258604317 2025-02-06 22:43:38   
2017   https://jo.opensooq.com/en/search/257306655 2025-02-06 23:33:19   
2067   https://jo.opensooq.com/en/search/258596595 2025-02-06 23:34:02   
2120   https://jo.opensooq.com/en/search/258583659 2025-02-06 23:34:47   
2128   https://jo.opensooq.com/en/search/258590863 2025-02-06 23:34:53   
2132   https://jo.opensooq.com/en/search/258590219 

#### Handling Missing Values in 'Price' Attribute

In [69]:
grouped = clean_df.groupby(["Model", "Year"])
for (model, year), group in grouped:
    if group['Price'].isna().any():
        indx = group['Price'][group['Price'].isna()].index  # Only missing values

        if group.shape[0] <= 2:
            car_make_value = group['Car Make'].iloc[0]
            mean_price = clean_df[clean_df['Car Make'] == car_make_value]['Price'].mean()

            # Update ONLY NaN values without affecting other columns
            clean_df.loc[indx, 'Price'] = clean_df.loc[indx, 'Price'].fillna(mean_price)

        elif group.shape[0] > 2:
            clean_df.loc[indx, 'Price'] = clean_df.loc[indx, 'Price'].fillna(group['Price'].mean())

print("Final shape of clean_df:", clean_df.shape)


Final shape of clean_df: (11247, 30)


In [70]:
clean_df[clean_df['Price'].isna()].shape[0]

11

In [71]:
indx2 = clean_df[clean_df['Price'].isna()].index

In [72]:
clean_df.drop(indx2,inplace =True)

In [73]:
clean_df.shape

(11236, 30)

In [74]:
# unhandeled outliers
clean_df[clean_df['Price']<=1001].shape[0]

106

In [75]:
# stronig indx of unhandeled outliers
groupless_outliers = clean_df[clean_df['Price']<=1001].index

In [76]:
# final count of missing values
clean_df.isna().sum()

URL                      0
Scraped_Time             0
Condition                0
Car Make                 0
Model                    0
Trim                     0
Year                     0
Kilometers               0
Body Type                0
Number of Seats          0
Fuel                     0
Transmission             0
Engine Size (cc)         0
Exterior Color           0
Interior Color           0
Regional Specs           0
Car License              0
Insurance                0
Car Customs              0
Body Condition           0
Paint                    0
Payment Method           0
City                     0
Neighborhood             0
Interior Options         0
Exterior Options         0
Technology Options       0
Price                    0
Battery Capacity      9322
Battery Range         9322
dtype: int64

In [77]:
# final shape of the dataset
clean_df.shape

(11236, 30)

In [78]:
#converting 'Price' from object to flaot
clean_df['Price'] = clean_df['Price'].astype(int)
clean_df['Price'].dtype

dtype('int32')

In [79]:
#convert 'Year' from object to integer
clean_df["Year"] = clean_df["Year"].astype(int)
clean_df['Year'].dtype

dtype('int32')

### Handling Duplicated Entities

*'URL'* atrribute used as unique identifier, where each link cotains unique *id* for each car ad, hence *URL's* are not redundant, and if there is a redundant *URL* it shuold be removed.
As shown in *Descovering Structure of the Data* section the number of duplicated rows in general are 6, while the number of dulpicates by *URL* are 76 and that is another reasons why using *URL* to handle duplicates

In [80]:
clean_df.duplicated(subset = ['URL']).sum()

67

In [81]:
clean_df.drop_duplicates(subset = ['URL'], inplace = True)

In [82]:
clean_df.shape

(11169, 30)

## Spliting & Exporting Data

In [83]:
electric_cars_df = clean_df[clean_df['Fuel'] == 'Electric'].copy()
electric_cars_df.drop(columns = ['Engine Size (cc)'], inplace = True)
electric_cars_df.shape

(1876, 29)

In [84]:
non_electric_cars_df = clean_df[(clean_df['Fuel'] == 'Gasoline')|
                                (clean_df['Fuel'] == 'Diesel')|
                                (clean_df['Fuel'] == 'Hybrid')|
                                (clean_df['Fuel'] == 'Plug-in - Hybrid')|
                                (clean_df['Fuel'] == 'Mild Hybrid')].copy()
non_electric_cars_df.drop(columns = ['Battery Range','Battery Capacity'], inplace = True)
non_electric_cars_df.shape

(9293, 28)

In [85]:
clean_df.to_csv("cars_data_cleaned.csv",index = False)

In [86]:
electric_cars_df.to_csv("Electric_cars_cleaned.csv",index = False)
non_electric_cars_df.to_csv("Non_Electric_cars_cleaned.csv",index = False)

In [87]:
electric_cars_df.isna().sum()

URL                   0
Scraped_Time          0
Condition             0
Car Make              0
Model                 0
Trim                  0
Year                  0
Kilometers            0
Body Type             0
Number of Seats       0
Fuel                  0
Transmission          0
Exterior Color        0
Interior Color        0
Regional Specs        0
Car License           0
Insurance             0
Car Customs           0
Body Condition        0
Paint                 0
Payment Method        0
City                  0
Neighborhood          0
Interior Options      0
Exterior Options      0
Technology Options    0
Price                 0
Battery Capacity      0
Battery Range         0
dtype: int64

In [88]:
non_electric_cars_df.isna().sum()

URL                   0
Scraped_Time          0
Condition             0
Car Make              0
Model                 0
Trim                  0
Year                  0
Kilometers            0
Body Type             0
Number of Seats       0
Fuel                  0
Transmission          0
Engine Size (cc)      0
Exterior Color        0
Interior Color        0
Regional Specs        0
Car License           0
Insurance             0
Car Customs           0
Body Condition        0
Paint                 0
Payment Method        0
City                  0
Neighborhood          0
Interior Options      0
Exterior Options      0
Technology Options    0
Price                 0
dtype: int64